### PR TITLE
[DM] [SASS-331] Improved page access restrictions

### PR DESCRIPTION
**Pull request by TheDeadlyPianist**

### CODE REVIEW

File: app/controllers/dividends/DividendsCYAController.scala [hunk 1]
```diff
@@ -127,4 +129,14 @@ class DividendsCYAController @Inject()(
     getSessionData[DividendsCheckYourAnswersModel](SessionValues.DIVIDENDS_CYA)
   }
 
+  private[dividends] def handleUnfinishedRedirect(cya: DividendsCheckYourAnswersModel, taxYear: Int): Result = {
+    DividendsCheckYourAnswersModel.unapply(cya).getOrElse(None, None, None, None) match {
+      case (None, None, None, None) => Redirect(controllers.dividends.routes.ReceiveUkDividendsController.show(taxYear))
```
Comment (robert-webb-hmrc) [inline app/controllers/dividends/DividendsCYAController.scala:134]: Manually testing this case redirected back to the Overview page, as per the case match on line 82. Will we ever see a cya result with All inputs being none? this is also the case for a true/false response on the UK Dividends page.

**Review by robert-webb-hmrc — CHANGES_REQUESTED** [review]
Couple of things to check. Can we also rebase this as please?

File: app/controllers/dividends/DividendsCYAController.scala (hunk 1 above)
Comment (TheDeadlyPianist) [reply app/controllers/dividends/DividendsCYAController.scala:134]: Deleting line 134 as it cannot happen IRL. Update to create CYA data on journey start, so the user doesn't get thrown to the Overview page if they try and hit the CYA page after pressing Yes.

File: app/controllers/interest/InterestCYAController.scala [hunk 2]
```diff
@@ -110,4 +111,16 @@ class InterestCYAController @Inject()(
     val event = AuditModel("CreateOrAmendInterestUpdate", "createOrAmendInterestUpdate", details)
     auditService.auditModel(event)
   }
+
+  private def handleUnfinishedRedirect(cya: InterestCYAModel, taxYear: Int): Future[Result] = {
+    Future.successful(
+      InterestCYAModel.unapply(cya).getOrElse(None, None, None, None) match {
+        case (None, None, None, None) => Redirect(controllers.interest.routes.UntaxedInterestController.show(taxYear))
+        case (Some(true), None, None, None) => Redirect(controllers.interest.routes.AccountsController.show(taxYear, InterestTaxTypes.UNTAXED))
```
[the remark this answers is not in the document]
Comment (TheDeadlyPianist) [reply app/controllers/interest/InterestCYAController.scala:119]: Update to send user to Amount Controller, not Account Controller.

File: app/controllers/interest/InterestCYAController.scala [hunk 3]
```diff
@@ -110,4 +111,16 @@ class InterestCYAController @Inject()(
...
+        case (Some(false), None, None, None) => Redirect(controllers.interest.routes.TaxedInterestController.show(taxYear))
+        case (Some(true), Some(_), None, None) => Redirect(controllers.interest.routes.TaxedInterestController.show(taxYear))
+        case _ => Redirect(controllers.interest.routes.AccountsController.show(taxYear, InterestTaxTypes.TAXED))
```
[the remark this answers is not in the document]
Comment (TheDeadlyPianist) [reply app/controllers/interest/InterestCYAController.scala:122]: See above

**Review by robert-webb-hmrc — APPROVED** [review]

**Pull request merged**